File: Python/Topics/Robot/Tablet.py
import qi
import json

class Tablet:
    tablet_service = None
    connection = None

    def __init__(self, tablet_service, connection):
        self.tablet_service = tablet_service
        self.connection = connection


    def interpretContent(self, message):
        content = message["content"]

        if content["tablet_command"] == "wifi_enableWifi":
            self.enableWifi()
        if content["tablet_command"] == "wifi_disableWifi":
            self.disableWifi()
        if content["tablet_command"] == "wifi_getWifiStatus":
            self.getWifiStatus()
        if content["tablet_command"] == "wifi_configureWifi":
            self.configureWifi(content["security"], content["network_name"], content["network_password"])
        if content["tablet_command"] == "wifi_connectWifi":
            self.connectWifi(content["network_name"])
        if content["tablet_command"] == "wifi_disconnectWifi":
            self.disconnectWifi()

        if content["tablet_command"] == "web_showWebView":
            self.showWebView(content["url"])
        if content["tablet_command"] == "web_hideWebView":
            self.hideWebView()
        if content["tablet_command"] == "web_loadApplication":
            self.loadApplication(content["application_name"])
        if content["tablet_command"] == "web_loadUrl":
            self.loadUrl(content["url"])
        if content["tablet_command"] == "web_reloadPage":
            self.reloadPage(content["bypass_cache"])

        if content["tablet_command"] == "video_playVideo":
            self.playVideo(content["url"])
        if content["tablet_command"] == "video_stopVideo":
            self.stopVideo()
        if content["tablet_command"] == "video_pauseVideo":
            self.pauseVideo()
        if content["tablet_command"] == "video_resumeVideo":
            self.resumeVideo()

        if content["tablet_command"] == "image_showImage":
            self.showImage(content["url"])
        if content["tablet_command"] == "image_hideImage":
            self.hideImage()

        if content["tablet_command"] == "dialog_showInputDialog":
            self.showInputDialog(content["type"], content["title"], content["ok_text"], content["cancel_text"])
        if content["tablet_command"] == "dialog_hideDialog":
            self.hideDialog()

        if content["tablet_command"] == "wake_up":
            self.tablet_service.wakeUp()
        if content["tablet_command"] == "sleep":
            self.tablet_service.goToSleep()

        if content["tablet_command"] == "lowLevel_closeTabletBrowser":
            self.closeTabletBrowser()
            return True
        return False

    def closeTabletBrowser(self):
        try:
            self.tablet_service._stopApk("jp.softbank.tabletbrowser")
        except RuntimeError:
            # A runtime error is expected, as the connection to the TabletService is lost, the moment the apk is stopped
            tablet_service = None
            pass

    def tabletNotificationConfirmation(self, ip, port):
        try:
            connection_url = "tcp://" + ip + ":" + str(port)
            app = qi.Application(["TabletModule", "--qi-url=" + connection_url])
            app.start()

            session = app.session
            behavior_mng_service = session.service("ALBehaviorManager")
            behavior_mng_service.runBehavior("notificationproj2-5d7801/behavior_1")
            self.tablet_service = session.service("ALTabletService")

            # Don't forget to disconnect the signal at the end
            signalID = 0

            # function called when the signal onTouchDown is triggered
            def callback(x, y):
                print "Notification has been confirmed"
                print "Touch coordinate are x: ", x, " y: ", y
                self.tablet_service.hideWebview()
                self.tablet_service.onTouchDown.disconnect(signalID)
                app.stop()

            # attach the callback function to onJSEvent signal
            signalID = self.tablet_service.onTouchDown.connect(callback)
            app.run()
        except Exception, e:
            print "Error was: ", e

    #region Wifi
    def enableWifi(self):
        self.tablet_service.enableWifi()

    def disableWifi(self):
        self.tablet_service.disableWifi()

    def getWifiStatus(self):
        print self.tablet_service.getWifiStatus()

    def configureWifi(self, security, network_name, network_password):
        self.tablet_service.configureWifi(security, network_name, network_password)

    def connectWifi(self, network_name):
        self.tablet_service.connectWifi(network_name)

    def disconnectWifi(self):
        self.tablet_service.disconnectWifi()
    #endregion

    #region Web View
    def showWebView(self, url):
        if url == "":
            self.tablet_service.showWebview()
        else:
            self.tablet_service.showWebview(url)

    def hideWebView(self):
        self.tablet_service.hideWebview()

    def loadApplication(self, application_name):
        self.tablet_service.loadApplication(application_name)

    def loadUrl(self, url):
        self.tablet_service.loadUrl(url)

    def reloadPage(self, bypassCache): # set to true in order to reload the current page by bypassing the local web cache
        self.tablet_service.reloadPage(bypassCache)
    #endregion

    #region Video Player
    def playVideo(self, url):
        self.tablet_service.playVideo(url)

    def stopVideo(self):
        self.tablet_service.stopVideo()

    def pauseVideo(self):
        self.tablet_service.pauseVideo()

    def resumeVideo(self):
        self.tablet_service.resumeVideo()
    #endregion

    #region Image
    def showImage(self, url):
        self.tablet_service.showImage(url)

    def hideImage(self):
        self.tablet_service.hideImage()
    #endregion

    #region Dialog
    def showInputDialog(self, type, title, ok_text, cancel_text):
        touchID = 0
        def touchdown(x,y):
            print str(x) + " " + str(y)
            if(x > 2000 and y > 500):
                self.tablet_service.onTouchDown.disconnect(touchID)

        #touchID = tablet_service.onTouchDown.connect(touchdown)
        touchID = self.tablet_service.onTouchMove.connect(touchdown)
        #drawing sites
        #https://www.tinyimage.com

        self.tablet_service.showInputDialog(type, title, ok_text, cancel_text)
        self.tablet_service.onInputText

        signalID = 0
        def callback(validation, input_text):
            x = ""
            if(validation == 1):
                print input_text
                x = {"input_text": input_text}
            if(validation == 0):
                print "Input canceled"
                x = {"input_text": "error::canceled"}
            y = json.dumps(x)
            self.connection.sendall(y.encode())

        signalID = self.tablet_service.onInputText.connect(callback)

    def hideDialog(self):
        self.tablet_service.hideDialog()
    #endregion

File: Unity/PePUT_Unity/Assets/Scripts/Backend/Tablet.cs
using System;
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class Tablet : MessageClient
{
    public struct TabletInput
    {
        public string inputText;
    }
    TabletInput currentInput;

    public Tablet(TCPClientTopic client) : base(client)
    {
    }

    #region Wifi
    public void EnableWifi()
    {
        TabletContent tabletContent = new TabletContent(TabletContent.TABLET_COMMAND.wifi_enableWifi);
        SendTabletContent(tabletContent);
    }

    public void DisableWifi()
    {
        TabletContent tabletContent = new TabletContent(TabletContent.TABLET_COMMAND.wifi_disableWifi);
        SendTabletContent(tabletContent);
    }

    public void ConsolePrintWifiStatus()
    {
        TabletContent tabletContent = new TabletContent(TabletContent.TABLET_COMMAND.wifi_getWifiStatus);
        SendTabletContent(tabletContent);
    }

    public void ConfigureWifi(string networkName, TabletContent.SECURITY security, string networkPassword)
    {
        TabletContent tabletContent = new TabletContent(TabletContent.TABLET_COMMAND.wifi_configureWifi, networkName, security, networkPassword);
        SendTabletContent(tabletContent);
    }

    public void ConnectWifi(string networkName)
    {
        TabletContent tabletContent = new TabletContent(TabletContent.TABLET_COMMAND.wifi_connectWifi, networkName);
        SendTabletContent(tabletContent);
    }

    public void DisconnectWifi()
    {
        TabletContent tabletContent = new TabletContent(TabletContent.TABLET_COMMAND.wifi_disconnectWifi);
        SendTabletContent(tabletContent);
    }
    #endregion

    #region Web View
    public void ShowWebView(string url)
    {
        TabletContent tabletContent = new TabletContent(TabletContent.TABLET_COMMAND.web_showWebView, url);
        SendTabletContent(tabletContent);
    }

    public void HideWebView()
    {
        TabletContent tabletContent = new TabletContent(TabletContent.TABLET_COMMAND.web_hideWebView);
        SendTabletContent(tabletContent);
    }

    public void LoadApplication(string applicationName)
    {
        TabletContent tabletContent = new TabletContent(TabletContent.TABLET_COMMAND.web_loadApplication, applicationName);
        SendTabletContent(tabletContent);
    }

    public void PrintInstalledApplications()
    {
        if (client == null)
        {
            Debug.LogError("CLIENT NOT INITIALIZED");
            return;
        }
        BehaviorContent behaviorContent = new BehaviorContent(BehaviorContent.BEHAVIOR_COMMAND.LIST);
        TCPContent tcpContent = new TCPContent(TCPContent.TOPIC.ROBOT, TCPContent.SUBTOPIC.ROBOT_BEHAVIOR_TOOL, 1, behaviorContent);
        client.SendMessage(tcpContent.toJSONMessage());
    }

    public void LoadUrl(string url)
    {
        TabletContent tabletContent = new TabletContent(TabletContent.TABLET_COMMAND.web_loadUrl, url);
        SendTabletContent(tabletContent);
    }

    public void ReloadWebPage(bool bypassCache = true)
    {
        TabletContent tabletContent = new TabletContent(TabletContent.TABLET_COMMAND.web_reloadPage, bypassCache);
        SendTabletContent(tabletContent);
    }
    #endregion

    #region Video
    public void PlayVideo(string url)
    {
        TabletContent tabletContent = new TabletContent(TabletContent.TABLET_COMMAND.video_playVideo, url);
        SendTabletContent(tabletContent);
    }

    public void StopVideo()
    {
        TabletContent tabletContent = new TabletContent(TabletContent.TABLET_COMMAND.video_stopVideo);
        SendTabletContent(tabletContent);
    }

    public void PauseVideo()
    {
        TabletContent tabletContent = new TabletContent(TabletContent.TABLET_COMMAND.video_pauseVideo);
        SendTabletContent(tabletContent);
    }

    public void ResumeVideo()
    {
        TabletContent tabletContent = new TabletContent(TabletContent.TABLET_COMMAND.video_resumeVideo);
        SendTabletContent(tabletContent);
    }
    #endregion

    #region Image
    public void ShowImage(string url)
    {
        TabletContent tabletContent = new TabletContent(TabletContent.TABLET_COMMAND.image_showImage, url);
        SendTabletContent(tabletContent);
    }

    public void HideImage()
    {
        TabletContent tabletContent = new TabletContent(TabletContent.TABLET_COMMAND.image_hideImage);
        SendTabletContent(tabletContent);
    }
    #endregion

    #region Input Dialog
    public void ShowInputDialog(TabletContent.TYPE type, string title, string okText = "ok", string cancelText = "cancel")
    {
        TabletContent tabletContent = new TabletContent(TabletContent.TABLET_COMMAND.dialog_showInputDialog, type, title, okText, cancelText);
        SendTabletContent(tabletContent);
    }

    public void HideInputDialog()
    {
        TabletContent tabletContent = new TabletContent(TabletContent.TABLET_COMMAND.dialog_hideDialog);
        SendTabletContent(tabletContent);
    }
    #endregion

    public void CloseTabletBrowser()
    {
        TabletContent tabletContent = new TabletContent(TabletContent.TABLET_COMMAND.lowLevel_closeTabletBrowser);
        SendTabletContent(tabletContent);
    }

    private void SendTabletContent(TabletContent tabletContent)
    {
        if (client == null)
        {
            Debug.LogError("CLIENT NOT INITIALIZED");
            return;
        }
        TCPContent tcpContent = new TCPContent(TCPContent.TOPIC.ROBOT, TCPContent.SUBTOPIC.ROBOT_TABLET, 1, tabletContent);
        client.SendMessage(tcpContent.toJSONMessage());
    }

    public override void ReceiveMessage<T>(ref T message_T)
    {
        string message = message_T.ToString();
        Debug.Log(message);

        try
        {
            JsonUtility.FromJson<TabletInput>(message);
        }
        catch (Exception e)
        {
            Debug.Log(e);
            return;
        }

        TabletInput input = JsonUtility.FromJson<TabletInput>(message);
        if (input.inputText == "error::canceled")
        {
            TabletInputCanceled();
            return;
        }
        Debug.Log(input.inputText);
        currentInput = input;

        NotifySubscribers(ref input);
    }

    private void TabletInputCanceled()
    {
        Debug.Log("Input was canceled");
    }

    protected override void NotifySubscribers<TabletInput>(ref TabletInput tabletInfo)
    {
        foreach(TCPMessageSubscriber subscriber in subscribers)
        {
            subscriber.ReceiveMessage(ref tabletInfo);
        }
    }
}
